lib/puppet/property.rb

```diff
@@ -205,6 +206,22 @@ def change_to_s(current_value, newvalue)
     end
   end
 
+  # Format the given string and conditionally redact the provided interpolation variables, depending on if
+  # this property is sensitive.
+  #
+  # @note Because the default implementation of #is_to_s returns the current value as-is, it doesn't necessarily
+  #   return a string. For the sake of sanity we just cast everything to a string for interpolation so we don't
+  #   introduce issues with unexpected property values.
+  #
+  # @see String#format
+  # @param fmt [String] The format string to interpolate.
+  # @param args [Array<String>] One or more strings to conditionally redact and interpolate into the format string.
+  #
+  # @return [String]
+  def format(fmt, *args)
+    fmt % args.map { |arg| @sensitive ? "[redacted]" : arg.to_s }
+  end
```

Comment:
I don't really like how much of sorting out "is" and "should" is left to the resource_harness in this... but that's really part of the existing design of the code, and I think this is as good as it's going to get without some major restructuring of the responsibilities of the resource_harness vs properties.